src/prefect/results.py

```diff
@@ -273,23 +273,30 @@ async def from_settings(
 
     @staticmethod
     async def resolve_storage_block(
-        result_storage: ResultStorage, client: "PrefectClient"
+        result_storage: ResultStorage,
+        client: "PrefectClient",
+        persist_result: bool = True,
     ) -> Tuple[uuid.UUID, WritableFileSystem]:
         """
         Resolve one of the valid `ResultStorage` input types into a saved block
         document id and an instance of the block.
         """
         if isinstance(result_storage, Block):
             storage_block = result_storage
-            storage_block_id = (
+
+            if storage_block._block_document_id is not None:
                 # Avoid saving the block if it already has an identifier assigned
-                storage_block._block_document_id
-                # TODO: Overwrite is true to avoid issues where the save collides with
-                #       a previously saved document with a matching hash
-                or await storage_block._save(
-                    is_anonymous=True, overwrite=True, client=client
-                )
-            )
+                storage_block_id = storage_block._block_document_id
+            else:
+                if persist_result:
+                    # TODO: Overwrite is true to avoid issues where the save collides with
+                    # a previously saved document with a matching hash
+                    storage_block_id = await storage_block._save(
+                        is_anonymous=True, overwrite=True, client=client
+                    )
+                else:
+                    # dummy UUID which should not matter
+                    storage_block_id = uuid.uuid4()
```

Comment:
During a sync review, we weren't able to determine what the downstream effects of returning a dummy UUID would be. We decided to return `None` instead and add assertions anywhere this `storage_block_id` is used that it isn't `None`. That way, if there are any downstream effects, they should be easier to debug if/when they occur.